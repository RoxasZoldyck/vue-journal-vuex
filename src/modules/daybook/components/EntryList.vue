<!-- Este componente contiene una lista de entradas y se encuentra dentro de DayBookLayout -->
<template>
    <div class="entry-list-container">
        <div class="px-2 pt-2">
            <input 
            type="text"
            class="form-control"
            placeholder="Buscar entradas"
            v-model="term"
            />
        </div>

        <!--Creamos el botón para agregar nuevas entradas-->
        <div class="mt-2 d-flex flex-column">
            <button class="btn btn-primary mx-3"
            @click="$router.push({name:'entry', params:{id:'new'}})">
                <i class="fa fa-plus-circle"></i>
                Nueva entrada
            </button>
        </div>


        <div class="entry-scrollarea">
            <!-- Primer ejercicio, llamar al getter getEntriesByTerm para que imprima las entradas configuradas en el state: -->
            <!-- Mi respuesta: -->
            <!-- También se puede hacer sin mapGetters con $store.getters['journalModule/getEntriesByTerm']-->
            <Entry
            v-for="entry in entriesByTerm"
            :key="entry.id"
            :entry="entry"> <!--Esto es importante, aquí con :entry se está mandando el objeto que contiene todas las entradas-->
            </Entry>
        </div>
    </div>
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters } from 'vuex';

export default {
    components:{
        Entry: defineAsyncComponent(()=>import('../components/Entry.vue')),
    },
    //Mi respuesta
    computed:{
        ...mapGetters('journalModule',['getEntriesByTerm']),
        //Esta función computada lo que hace es regresar getEntriesByTerm pero con el argumento del term, el cuál está vinculado al input Buscar entradas por medio de un v-model
        entriesByTerm(){
            return this.getEntriesByTerm(this.term)
        }
    },
    data(){
        return {
            term:''
        }
    }
}
</script>

<style lang="scss" scoped>

.entry-list-container{
    border-right: 1px solid #2c3e50;
    height: calc(100vh - 56px); //100vh es la altura total, y se le resta 56 px
}

.entry-scrollarea{
    height: calc(100vh - 110px);//100vh es la altura total, y se le resta 110 px
    overflow: scroll; //Habilitamos el scroll
}
</style>