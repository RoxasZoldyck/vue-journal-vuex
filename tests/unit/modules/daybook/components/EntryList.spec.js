import { createStore } from "vuex"
import journalModule from "@/modules/daybook/store/journal"
import { journalState } from "../../../mock-data/test-journal-state"
import { shallowMount } from "@vue/test-utils"
import EntryList from "@/modules/daybook/components/EntryList.vue"


const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journalModule: {
                ...journalModule, //Desestructuramos todo lo que journalModule contenga
                state: { ...initialState } //Sobreescribimos el state con lo que haya en initialState
            }
        }
    })

describe('Pruebas en el EntryList', () => {
    const store = createVuexStore(journalState) //Creamos nuestro store virtual
    const mockRouter = {
        push: jest.fn()
    } //Creamos el router virtual, este básicamente se encarga de que se cumplan $router

    let wrapper

    //Hacemos esto para que el wrapper se reinicie cada que se ejecute una prueba
    beforeEach(() => {
        jest.clearAllMocks() //Limpiamos todos los mocks
        wrapper = shallowMount(EntryList, {
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('debe de llamar el getEntriesByTerm sin termino y mostrar 2 entradas', () => {

        expect(wrapper.findAll('entry-stub').length).toBe(2) //Validamos que haya sólo 2 objetos de entry-stub
        expect(wrapper.html()).toMatchSnapshot() //Validamos la snapshot

    })

    test('debe de llamar el getEntriesByTerm y filtrar las entradas', async () => {

        const input = wrapper.find('input') //Buscamos el input
        await input.setValue('Mariana') //Colocamos en el input lo que queremos buscar

        expect(wrapper.findAll('entry-stub').length).toBe(1) //Sólo debe de regresar una entrada

    })

    test('el boton de nuevo debe de redireccionar a /new', () => {

        wrapper.find('button').trigger('click') //Presionamos el butón y mandamos un click virtual

        expect(mockRouter.push)
            .toHaveBeenCalledWith({ name: 'entry', params: { id: 'new' } }) //Validamos que el router se llame con el objeto { name: 'entry', params: { id: 'new' } }, el cuál es el que te envía a la ruta entry

    })
})