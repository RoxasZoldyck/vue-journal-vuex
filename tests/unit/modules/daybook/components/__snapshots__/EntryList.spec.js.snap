// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryList debe de llamar el getEntriesByTerm sin termino y mostrar 2 entradas 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar entradas"></div>
  <!--Creamos el botón para agregar nuevas entradas-->
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <!-- Primer ejercicio, llamar al getter getEntriesByTerm para que imprima las entradas configuradas en el state: -->
    <!-- Mi respuesta: -->
    <!-- También se puede hacer sin mapGetters con $store.getters['journalModule/getEntriesByTerm']-->
    <entry-stub entry="[object Object]"></entry-stub>
    <entry-stub entry="[object Object]"></entry-stub>
  </div>
</div>
`;
